text/0715-construct-upgrade-support.md

```diff
@@ -0,0 +1,514 @@
+performance optimizations, and reliability improvements. To maintain backward compatibility
+while enabling forward progress, the CDK team uses structured versioning strategies for critical updates.
+
+A common pattern involves releasing versioned constructs or modules (e.g., V2 variants)
+that coexist with their predecessors. For example:
+
+* The `@aws-cdk/aws-ec2-alpha` alpha module introduces the modernized `VpcV2` construct,
+designed to succeed the original Vpc in `aws-cdk-lib/aws-ec2`
+* The `TableV2` construct in `aws-cdk-lib/aws-dynamodb` replaces legacy Table implementations
+while retaining API compatibility
+
+These updates unlock enhanced capabilities. Migration between versions, however, often requires
+code changes that could trigger CloudFormation resource replacements (e.g., NAT gateways being
+recreated, DynamoDB global tables being recreated). For stateful production resources, this can
+lead to availability risks and operational complexity.
+
+To address this, the CDK introduces Migration Support — a combination of guided workflows and
+automated safeguards designed to ensure safe adoption of new constructs while preserving existing
+infrastructure. This framework enables developers to upgrade seamlessly with minimal risk to
+production environments.
+
+### How it works
+
+The CDK team develops a construct upgrade framework designed to simplify adoption while minimizing
+risks. This approach combines guided documentation, automated validations, and deployment safeguards
+to ensure a controlled transition.
+
+In a high-level overview, for each supported migration, the construct upgrade framework provides:
+
+- A comprehensive, step-by-step migration guide enabling users to transition from legacy constructs/modules
+to their modern equivalents.
+- A CDK CLI command `cdk construct-upgrade` to validate code changes and execute deployments safely.
+
+Note that each migration guide follows a standardized structure that includes dependency changes, behavioral
+analysis, risk assessments, and rollback instructions. This ensures consistency, clarity, and safety across
+all migrations. See [Appendix A](#appendix-a-migration-guide) for a detailed breakdown of the proposed migration
+guide format.
+
+With the initial release of the CDK construct upgrade tool, developers use the migration guide to
+update their CDK code, replacing legacy constructs (e.g., Table) with modern equivalents (e.g., TableV2).
+Developers have the flexibility to implement changes either manually or with the assistance of AI-powered
+code transformation tools (e.g., LLMs).
+
+After updating the code to use the new construct, the CDK CLI command `cdk construct-upgrade` validates the
+existing CDK stack and environment, ensuring they are ready for migration. The tool flags any warnings or
+blockers, such as CloudFormation drift, and cross-compares the old and new CloudFormation templates.
```

Comment:
Can you be specific about the types of changes we would flag here?

Does this basically amount to `cdk diff` shows no diff (or at least: no replacement)... perhaps that diff requirement is softened a little. Perhaps we accept some changes that we know are safe.

If so, wouldn't these kinds of safety checks make sense as part of `cdk diff` proper as well? They don't seem to need to be locked behind a new subcommand.